### ISSUE TITLE
Debian: MQTT Not detecting disconnects + Loss of messages

### ISSUE THREAD

**Issue by dsjh** [Azure/azure-iot-sdk-node#132]
# Context
- **OS and version used:** Debian Jessie (8) | 3.16.0-4-amd64 #1 SMP Debian 3.16.43-2+deb8u2 (2017-06-26) x86_64 GNU/Linux 
- **Node.js version:** 4.32
- **npm version:** 2.14.12
- **list of installed packages:** azure-iot-device-mqtt@1.2.1 node_modules/azure-iot-device-mqtt
├── azure-iot-common@1.2.1
├── debug@3.1.0 (ms@2.0.0)
├── azure-iot-mqtt-base@1.2.1 (mqtt@2.13.1)
└── machina@2.0.0 (lodash@3.10.1)
azure-iot-device@1.2.0 node_modules/azure-iot-device
├── debug@3.1.0 (ms@2.0.0)
├── traverse@0.6.6
├── azure-iot-common@1.2.1
├── azure-iot-http-base@1.2.0 (uuid@2.0.3, debug@2.6.9, azure-iot-common@1.2.0)
├── azure-storage@2.6.0 (extend@1.2.1, underscore@1.8.3, xmlbuilder@0.4.3, browserify-mime@1.2.9, validator@3.35.0, uuid@3.1.0, readable-stream@2.0.6, md5.js@1.3.4, json-edm-parser@0.1.2, xml2js@0.2.7)
├── machina@2.0.0 (lodash@3.10.1)
└── lodash@4.17.4
azure-iot-device-amqp@1.2.1 node_modules/azure-iot-device-amqp
├── uuid@3.1.0
├── azure-iot-common@1.2.1
├── debug@3.1.0 (ms@2.0.0)
├── machina@2.0.0 (lodash@3.10.1)
├── async@2.5.0 (lodash@4.17.4)
└── azure-iot-amqp-base@1.2.1 (amqp10-transport-ws@0.0.5, bluebird@3.5.1, amqp10@3.5.3)
- **cloned repo: N/A

# Description of the issue:
We are running the simple device code (here: https://github.com/Azure/azure-iot-sdk-node/blob/master/device/samples/simple_sample_device.js). When the internet connection drops the C2D messages are not getting queued (they are getting black-holed) and D2C messages are also getting lost (with no errors).

The client on `disconnect` method does not trigger and connection remains offline and outbound messages are still generated but have no send status. Inbound messages are black-holed and device explorer shows disconnected after about 15 seconds.

After 15 seconds C2D messages appear in the queue and the client still thinks it's connected.

Windows behaves the same (in that in the 15 second period, it loses messages). However Windows does detect a disconnection client-site and re-establishes the connection automatically.

# Code sample exhibiting the issue:
https://github.com/Azure/azure-iot-sdk-node/blob/master/device/samples/simple_sample_device.js

# Console log of the issue:
The client shows client connected and messages continue to be generated.

#When connected messages generated are:
```
Sending message: {"deviceId":"myFirstDevice","windSpeed":12.937402627430856,"temperature":24.324122960679233,"humidity":68.76592649612576}
send status: MessageEnqueued
Sending message: {"deviceId":"myFirstDevice","windSpeed":12.199973709881306,"temperature":24.004879547283053,"humidity":76.89963116776198}
send status: MessageEnqueued
```

#When disconnected messages generated are:
```
Sending message: {"deviceId":"myFirstDevice","windSpeed":12.024511671625078,"temperature":25.682738567702472,"humidity":60.161663251928985}
Sending message: {"deviceId":"myFirstDevice","windSpeed":12.770792583934963,"temperature":21.66648893384263,"humidity":74.64088747743517}
```

**Comment by pierreca** [comment]
Regarding not firing the `disconnect` event: we unfortunately have to play within the limitations of the platform. Unlike windows (which you noted) Linux doesn't kill sockets when the network connection goes down. since the socket is the lowest level object we have, we don't either. To solve that we started introducing a keepalive mechanism - in the MQTT package, we send a keepalive ping every 3 minutes which means a disconnect event will be fired after 6 minutes. We haven't gotten to implementing an API to change this keepalive value yet, nor do we have gotten to implementing this in AMQP but both these items are in our backlog.

Regarding the messages being lost - I'm not sure I understand what you're doing, would you mind clarify the following?

- when you say "C2D messages are not being queued": what are you running? `iothub-explorer` or the `azure-iothub` service SDK? are you running on the same local machine?
- "after 15 seconds C2D messages appear in the queue" -> which queue are you talking about?
- Regarding D2C Messages: they are being queued and the SDK will send them when the connection is reestablished. if the connection is not reestablished, they will be timed out after 4 minutes and the callback for `sendEvent` should be called. That's because we introduced in version 1.2.0 a retry mechanism. you can modify this behavior using the `setRetryPolicy` API.

**Comment by pierreca** [comment]
@dsjh 

re-reading the [corresponding stack overflow question](https://stackoverflow.com/questions/47077843/azure-iot-hub-mqtt-messages-lost-on-internet-drop)
I think I have a slightly better understanding of the question, because you talk about Device Explorer there (and not here).

- you have a device, either connected in MQTT or AMQP, one one side, and device explorer on the other side
- then you disconnect the device
- then you are trying to send C2D messages to the device (and the queue you are talking about is the internal C2D queue for the specific device, on IoT Hub)
- in parallel, you let the device send D2C messages.

that is my current understanding of your experiment. can you please confirm if that's correct or not?

if that is correct:
- D2C messages on the device are held by the SDK and will be sent when the connection is reestablished, as long as it's re-established within 4 minutes (default timeout). You can modify the retry mechanism by using the `setRetryPolicy` API of the Client object in the device SDK. What behavior would you like to see? 

- C2D messages: IoT hub relies on some keep alive mechanism and the presence of a socket to determine the connected status of the device - like any socket it can fail because of the network and it can take a little bit of time to detect this on the IoT Hub side. using AMQP you can use [feedback](https://docs.microsoft.com/en-us/azure/iot-hub/iot-hub-devguide-messages-c2d) messages to verify if the message was delivered and accepted by the device correctly or use [direct methods](https://docs.microsoft.com/en-us/azure/iot-hub/iot-hub-devguide-direct-methods) which provide a "synchronous" way of sending a payload to the device and can be made to fail immediately if the device isn't connected (or retry for a time). If you want to send a payload to the device regardless of whether it's connected or not you could also look at the [device twin](https://docs.microsoft.com/en-us/azure/iot-hub/iot-hub-devguide-device-twins) feature

With regards to detecting disconnections on the device side, I stand by my previous comment, we have a few improvements planned in the backlog, but have to deal with limitations of whatever platform we're running on.

With these explanations, would you care to elaborate on the behavior you would like to see and where you think the SDK (or the service) has an issue you would like to see resolved?

**Comment by dsjh** [comment]
... is my current understanding of your experiment. can you please confirm if that's correct or not?
> This is exactly the setup of the experiment

With these explanations, would you care to elaborate on the behavior you would like to see and where you think the SDK (or the service) has an issue you would like to see resolved?
> I understand that the behaviour is down the limitations of the platform. Our main focus is synchronising data to devices when either online or offline and I will investigate the number of suggestions you have.

> The customisable keep-alive would a great feature for people who would like to detect disconnections sooner. Apart from this I don't see any issues with the SDK.

Thanks for your comprehensive response no the behaviour of the software and suggestions.
-David

**Comment by dsjh** [comment]
Hi Pierre,

We have done some further testing today with both AMQP and MQTT and it appears that we can't get any type of disconnect notification (using the sample code). We have tried dropping the connection for 10 seconds to 20 minutes. 

Since this is the case we can't reliably send to a Linux device from the cloud as Direct method will fail (as device thinks its still connected) and feedback messages will never reach the cloud.

Device Twins are not feasible because of the type of data we are sending and even if feasible the device still thinks it's connected.

In simple terms we have a VM that we disable the Network Adapter on for testing. Are we doing something wrong or is there another way to determine connectivity.

Any advice would be greatly appreciated.

Kind regards,
David

**Comment by pierreca** [comment]
@dsjh 2 separate things:

> We have done some further testing today with both AMQP and MQTT and it appears that we can't get any type of disconnect notification (using the sample code). We have tried dropping the connection for 10 seconds to 20 minutes.

That worries me, it should detect it. **Can you share what VM type you are using and how you disable the connection?** I would like to try and reproduce the problem. 
FYI, we have some network testing samples that were based on docker: https://github.com/Azure/azure-iot-sdk-node/tree/master/network_e2e

> Since this is the case we can't reliably send to a Linux device from the cloud as Direct method will fail (as device thinks its still connected) and feedback messages will never reach the cloud.

Both C2D messages and device methods can ensure reliable message delivery to the device, slightly differently:
- C2D messages should use AMQP and request a "full" acknowledge from the device ("real" full acknowledge is only available in AMQP and HTTP, not MQTT) - but on the device side you must ensure idempotence, in other words your device code should behave properly if the same message is delivered twice, just in case the message was delivered once but the device didn't have the time to send the feedback before disconnection (in that case the message would be re-delivered on the next connection)

- Device Methods: you have 2 timeouts in the device method parameters structure to "adjust" things: `connectTimeoutInSeconds` (which waits for the device to have a valid connection with the hub) and `responseTimeoutInSeconds` (which waits for the device to send a response). Even if `connectTimeoutInSeconds` might not fire because IOT Hub believes the device is connected when it's not, `responseTimeoutInSeconds` will fire, if it is faster than the hub discovering the device is disconnected (methods calls do not survive disconnection: in other words, if a device is disconnected while a method has been called and a response hasn't been received, and it is detected by the hub, the method will be considered failed). 

The difference between the 2 options is that C2D messages are buffered in a queue on the hub (and you rely on feedback coming back later: meaning you have to be able to "fire and forget", finetune message timeout, and keep track of messages for which you have received feedback or not) whereas device methods are more "synchronous" (they will fail within the connect/response timeout) but your service needs to keep retrying whereas with C2D, the hub does that for you (as long as feedback is not received the hub will retry).

**Comment by pierreca** [comment]
@dsjh I may have found why the disconnect event is not emitted when using MQTT:
https://github.com/mqttjs/MQTT.js/issues/710

not sure if we're using the mqtt.js library in an unanticipated way or if there's a bug, but in either case that'd explain your observation when trying to drop the connection. I'll follow up and either modify our code or offer a PR there to fix this asap.

**Comment by bmrpatel** [comment]
Thanks @dsjh and @pierreca, I am experiencing the same problem. It not firing callback for the `sendEvent `even after 4 minutes, once disconnected. Also `disconnect` event is not fired either. 

While I was trying to come up with my own strategy to reconnect. I noticed that on reconnects, it is throwing errors below:
```
(node:7764) MaxListenersExceededWarning: Possible EventEmitter memory leak detected. 11 close listeners added. Use emitter.setMaxListeners() to increase limit
2017-11-04T14:59:53.768Z - warn: [0000001] PROCESS warning: { MaxListenersExceededWarning: Possible EventEmitter memory leak detected. 11 close listeners added. Use emitter.setMaxListeners() to increase limit
    at _addListener (events.js:281:19)
    at MqttBase.addListener (events.js:298:10)
    at D:\Projects\test\node_modules\azure-iot-device-mqtt\lib\mqtt.js:211:45
    at constructor._onEnter (D:\Projects\test\node_modules\azure-iot-mqtt-base\lib\mqtt_base.js:75:25)
    at constructor.transition (D:\Projects\test\node_modules\machina\lib\machina.js:653:28)
    at constructor.Fsm.(anonymous function) [as transition] (D:\Projects\test\node_modules\machina\lib\machina.js:466:63)
    at D:\Projects\test\node_modules\azure-iot-mqtt-base\lib\mqtt_base.js:64:44
    at MqttClient.<anonymous> (D:\Projects\test\node_modules\azure-iot-mqtt-base\lib\mqtt_base.js:240:13)
    at emitOne (events.js:120:20)
    at MqttClient.emit (events.js:210:7)
  name: 'MaxListenersExceededWarning',
```
 Is this mqtt bug?

Thanks

**Comment by pierreca** [comment]
@bmrpatel The EventEmitter leak looks like a separate issue

**Comment by wmmihaa** [comment]
I get the same behavior as @bmrpatel

**Comment by bmrpatel** [comment]
Thanks @pierreca, should I open new issue?

**Comment by pierreca** [comment]
@bmrpatel yes please, ideally with the console log and a sample code exhibiting the issue?

**Comment by dsjh** [comment]
Hi Pierre,
The issue of not detecting connection lost (waited over 30 mins) also happens with AMQP. We are using VMWare ESxi and we disable the network interface by un-ticking connect in the VM Network options setting. What can we do to help diagnose this further? Are there any logs we can collect?

**Comment by pierreca** [comment]
@dsjh sure: you can set the `DEBUG` environment variable to `azure-iot*` at first and if it doesn't show anything useful you can set it to just `*` to see them interleaved with the `amqp10` library.

**Comment by pierreca** [comment]
Quick update: I submitted a PR to the MQTT library repository and created a PR here too to change our way of handling the close event, so that no matter what the MQTT folks decide to do, we should be good. 

The AMQP and event handler leaks require different issues though.

**Comment by pierreca** [comment]
@bmrpatel @wmmihaa found the event handler leak and corrected it in the same PR - waiting for reviews before I check this is but this will make it to the next release

@dsjh regarding AMQP not detecting disconnects I have to dig into how keepalive works with the amqp10 library but i have a hunch it's also a keepalive thing.

**Comment by dsjh** [comment]
@pierreca after further investigation we discovered the libraries behave mostly as expected by blocking the firewall rather than disabling the network interface on VMware. We are now using AMQP and can see the following error when disconnect are detected (roughly 5 mins after blocking traffic)

```
(node) warning: possible EventEmitter memory leak detected. 11 disconnected listeners added. Use emitter.setMaxListeners() to increase limit.
Trace
    at AMQPClient.addListener (events.js:239:17)
    at Amqp._fsm.machina.Fsm.states.connecting._onEnter (/var/anpr/node_modules/azure-iot-device-amqp/node_modules/azure-iot-amqp-base/lib/amqp.js:163:37)
    at _.extend.transition (/var/anpr/node_modules/azure-iot-device-amqp/node_modules/machina/lib/machina.js:653:28)
    at Fsm.(anonymous function) [as transition] (/var/anpr/node_modules/azure-iot-device-amqp/node_modules/machina/lib/machina.js:466:63)
    at Amqp._fsm.machina.Fsm.states.disconnected.connect (/var/anpr/node_modules/azure-iot-device-amqp/node_modules/azure-iot-amqp-base/lib/amqp.js:109:36)
    at _.extend.handle (/var/anpr/node_modules/azure-iot-device-amqp/node_modules/machina/lib/machina.js:613:25)
    at Fsm.(anonymous function) [as handle] (/var/anpr/node_modules/azure-iot-device-amqp/node_modules/machina/lib/machina.js:466:63)
    at Amqp.connect (/var/anpr/node_modules/azure-iot-device-amqp/node_modules/azure-iot-amqp-base/lib/amqp.js:397:19)
    at Amqp._this._fsm.machina.Fsm.states.connecting._onEnter (/var/anpr/node_modules/azure-iot-device-amqp/lib/amqp.js:216:37)
    at _.extend.transition (/var/anpr/node_modules/azure-iot-device-amqp/node_modules/machina/lib/machina.js:653:28)
    at Fsm.(anonymous function) [as transition] (/var/anpr/node_modules/azure-iot-device-amqp/node_modules/machina/lib/machina.js:466:63)
    at Amqp._this._fsm.machina.Fsm.states.disconnected.connect (/var/anpr/node_modules/azure-iot-device-amqp/lib/amqp.js:118:77)
    at _.extend.handle (/var/anpr/node_modules/azure-iot-device-amqp/node_modules/machina/lib/machina.js:613:25)
    at Fsm.(anonymous function) [as handle] (/var/anpr/node_modules/azure-iot-device-amqp/node_modules/machina/lib/machina.js:466:63)
    at Amqp._this._fsm.machina.Fsm.states.disconnected.enableMethods (/var/anpr/node_modules/azure-iot-device-amqp/lib/amqp.js:180:36)
    at _.extend.handle (/var/anpr/node_modules/azure-iot-device-amqp/node_modules/machina/lib/machina.js:613:25)
```

Should I raise this as a new issue? It sounds very similar to that reported by @bmrpatel  and @wmmihaa just with AMQP instead of MQTT.

**Comment by pierreca** [comment]
sure - a new issue would be nice.
The fix for the MQTT event leak that @bmrpatel and @wmmihaa is waiting for approval in a pull request, and the EventEmitter that is leaking is not the same (different name) so it's not the same thing.

**Comment by pierreca** [comment]
Flagging as bug and fix-checked-in with regards to the original issue. the PR for the fix includes a fix for the EventEmitter leak that @bmrpatel reported too so no need for a separate issue here.

@dsjh for the AMQP EventEmitter leak problem I'd be grateful if you could open another issue with console logs and maybe a quick recap of the experiment that causes this.

**Comment by snobu** [comment]
Which branch contains the checked in fix?

**Comment by Kapster17** [comment]
How do we get the fix for MQTT Not detecting disconnects ?
we are @ 1.2.1

**Comment by pierreca** [comment]
we're looking at releasing new packages today or tomorrow, in the meantime you can clone the master branch of this repository.

**Comment by pierreca** [comment]
A new release including the fix has been published:
https://github.com/Azure/azure-iot-sdk-node/releases/tag/2017-11-15

Thanks again for your help!
